test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetProjectManagerServiceTests.cs

```diff
@@ -536,6 +541,757 @@ public async Task GetUpdateActionsAsync_WithPackageReferenceProject_WhenUpdating
+
+                TelemetryActivity.NuGetTelemetryService = new NuGetVSTelemetryService(telemetrySession.Object);
+
+                // Prepare: Force a nuget Restore
+                var command = new RestoreCommand(request);
+                // Force writing project.assets.json
+                var result = await command.ExecuteAsync();
+                await result.CommitAsync(logger, CancellationToken.None);
+
+                Assert.True(result.Success);
+                Assert.True(File.Exists(pajFilepath));
+
+                // Act I
+                var topPackages = await _projectManager.GetTransitivePackageOriginAsync(
+                    new PackageIdentity("PackageB", new NuGetVersion(1, 0, 0)),
+                    projectId,
+                    CancellationToken.None);
+
+                // Verify I
+                Assert.NotNull(topPackages);
+                Assert.NotEmpty(topPackages);
+                Assert.Equal(4, topPackages.Count); // 3 fw/RID pairs + 1 fw/null-RID pair = 4 elements
+                var fwRidEntry = topPackages.First();
+                Assert.Equal(1, fwRidEntry.Value.Count); // only one top dependency
+                var dep = fwRidEntry.Value.First();
+                Assert.Equal("packageA", dep.Identity.Id);
+                Assert.Equal(new NuGetVersion("2.0.0"), dep.Identity.Version);
+
+                // Act II
+                var topPackages2 = await _projectManager.GetTransitivePackageOriginAsync(
+                    new PackageIdentity("packageD", new NuGetVersion(0, 1, 1)),
+                    projectId,
+                    CancellationToken.None);
+
+                // Verify II
+                Assert.NotNull(topPackages2);
+                Assert.Equal(8, topPackages2.Count); // multitargeting: 2 keys * 3 RIDs + 2 fw/null-RID pair = 8 elements
+                string nullKey = null;
+                var keyNetFx = Tuple.Create(NuGetFramework.Parse("net472"), nullKey);
+                var keyNetCore = Tuple.Create(NuGetFramework.Parse("net5.0"), nullKey);
+                Assert.Collection(topPackages2[keyNetCore],
+                    x => AssertElement(x, "packageA", "2.0.0"),
+                    x => AssertElement(x, "packageX", "3.0.0"));
+                Assert.Collection(topPackages2[keyNetFx],
+                    x => AssertElement(x, "packageX", "3.0.0"));
+
+                // Act III: Unknown transitive dependency
+                var topPackages3 = await _projectManager.GetTransitivePackageOriginAsync(
+                    new PackageIdentity("abc", new NuGetVersion(0, 1, 1)),
+                    projectId,
+                    CancellationToken.None);
+
+                Assert.Empty(topPackages3);
+            }
+        }
+
+        private void AssertElement(IPackageReferenceContextInfo pkg, string id, string version)
+        {
+            Assert.Equal(id, pkg.Identity.Id);
+            Assert.Equal(NuGetVersion.Parse(version), pkg.Identity.Version);
+        }
```

Comment:
Could we add a UT that validates the cancel operation? Also, we should have UTs for invalid input.

Reply:
Added test. See GetInstalledAndTransitivePackagesAsync_WithCancellationToken_ThrowsAsync